src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -6,128 +6,231 @@
-    /**
-     * Registers (adds) a Parameter to the registry. Since the unique pointer is moved
-     * into the registry, the pointer may not be accessed by the caller after this
-     * function has been called.
-     *
-     * @param parameter A unique pointer to the Parameter to add. This pointer may not
-     * be accessed by the caller after this function has been called.
-     */
-    static void registerParameter(std::unique_ptr<Parameter<T>> parameter)
-    {
-        Parameter<T>::getMutableRegistry().emplace_back(std::move(parameter));
-    }
-
-    /**
-     * Updates all the Parameters of type T with the latest values from the ROS
-     * Parameter Server
-     */
-    static void updateAllParametersFromROSParameterServer()
-    {
-        for (const auto& p : Parameter<T>::getRegistry())
+    public:
+        /**
+         * Constructs a new Parameter
+         *
+         * @param parameter_name The name of the parameter used by dynamic_reconfigure
+         * @param default_value The default value for this parameter
+         */
+        explicit Parameter<T>(const std::string& parameter_name, T default_value)
+        {
+            this->name_ = parameter_name;
+            this->value_ = default_value;
+	    this->internal_param_ = std::make_shared<Parameter<T>>(*this);
+
+            Parameter<T>::registerParameter(this->internal_param_);
+        }
+
+        /**
+         * Returns the global path in the ROS parameter server where this parameter is stored
+         *
+         * @return the global path in the ROS parameter sever where this parameter is stored
+         */
+        const std::string getROSParameterPath() const
+        {
+            return NamespaceForParameters + "/" + name();
+        }
+
+        /**
+         * Returns the value of this parameter
+         *
+         * @return the value of this parameter
+         */
+        const T value() const
+        {
+	    if (internal_param_.get() != nullptr){
+            	return this->internal_param_->value();
+	    } else {
+		return this->value_;
+	    }
+        }
+        /**
+         * Returns the name of this parameter
+         *
+         * @return the name of this parameter
+         */
+        const std::string name() const
+        {
+            return name_;
+        }
+
+        /**
+         * Updates the value of this Parameter with the value from the ROS
+         * Parameter Server
+         */
+        void updateValueFromROSParameterServer()
+        {
+            ros::param::get(getROSParameterPath(), this->value_);
+        }
+
+        /**
+         * Sets the parameter value in the ROS Parameter Server to the new value.
+         *
+         * @param new_value The new value to be set
+         */
+        void setValueInROSParameterServer(T new_value)
         {
-            p->updateValueFromROSParameterServer();
+	    //dynamic reconfigure takes control of setting params 
+	    //ros::param::set(getROSParameterPath(), new_value);
```

Comment:
Just to confirm, this isn't possible for us to do as well? (ie. we can't do something like making a call to the dynamic reconfigure server from here?) If it's not possible, we should probably remove this function.

@MathewMacDougall - just to double check, do things still work as you had planned them to (or closely enough) with this function removed?

Reply:
We actually can and the GUI will update, but not sure if it’s desired behavior because on the GUI any value set will be overridden by code in which case external configuration doesn’t really work, we could however, for Params that don’t need to be configured through the GUI, have some sort of flag internally and enable this function

Reply:
Yes things will still work without the ability to set parameters from within the code. The logic code will typically only read parameters. The parameters will be set either by the Dynamic_Reconfigure GUI, or our own Visualizer.

@akhilveeraghanta which GUI are you referring to? The DynamicReconfigure one? Either way I think we should keep this simple and not add any flags or conditional behavior for updating. We will be fine without this function.

Reply:
fixed